R/generate.R

```diff
@@ -126,12 +126,14 @@ sanitize_generation_type <- function(x) {
 }
 
 # Ensure that the supplied type matches what would be assumed from input
-compare_type_vs_auto_type <- function(type, auto_type) {
+compare_type_vs_auto_type <- function(type, auto_type, x) {
   if(is.null(auto_type)) {
     return(type)
   }
   if (auto_type != type &&
-      (any(!c(auto_type, type) %in% c("draw", "simulate")))) {
+      (any(!c(auto_type, type) %in% c("draw", "simulate"))) &&
+      is_hypothesized(x) ||
+      (type == "bootstrap" && auto_type == "draw" && is_hypothesized(x))) {
```

Comment:
Can you clarify the logic here? It looks right but I want to make sure I'm following it correctly.

Reply:
Ah, thank you for the nudge—realized that the condition after the OR was encompassed by the one preceding. Fixed now!

Of course. The [lines](https://github.com/tidymodels/infer/blob/aa59b0b0a4aeba70b292650439c3a97629a85035/R/generate.R#L133-L135) raise a warning when:

- The generated `auto_type` and supplied `type` are meaningfully different: they don't match and it's not just switching `"draw"` out for `"simulate"` or vice versa. AND
- We're in the context of hypothesis testing.

Examples of current functionality: 

``` r
library(infer)
  
# expected and given "bootstrap" (hypothesized)
gss %>%
  specify(response = hours) %>%
  hypothesize(null = "point", mu = 40) %>%
  generate(reps = 200, type = "bootstrap")
#> Response: hours (numeric)
#> Null Hypothesis: point
#> # A tibble: 100,000 × 2
#> # Groups:   replicate [200]
#>    replicate hours
#>        <int> <dbl>
#>  1         1  48.6
#>  2         1  41.6
#>  3         1  38.6
#>  4         1  87.6
#>  5         1  22.6
#>  6         1  38.6
#>  7         1  10.6
#>  8         1  11.6
#>  9         1  58.6
#> 10         1  76.6
#> # … with 99,990 more rows

# expected and given "bootstrap" (no hypothesis)
gss %>%
  specify(response = hours) %>%
  # hypothesize(null = "point", mu = 40) %>%
  generate(reps = 200, type = "bootstrap")
#> Response: hours (numeric)
#> # A tibble: 100,000 × 2
#> # Groups:   replicate [200]
#>    replicate hours
#>        <int> <dbl>
#>  1         1    35
#>  2         1    40
#>  3         1    15
#>  4         1    40
#>  5         1    40
#>  6         1    55
#>  7         1    25
#>  8         1     4
#>  9         1    30
#> 10         1    45
#> # … with 99,990 more rows

# expected bootstrap, given draw (hypothesized)
gss %>%
  specify(response = hours) %>%
  hypothesize(null = "point", mu = 40) %>%
  generate(reps = 200, type = "draw")
#> Warning: You have given `type = "draw"`, but `type` is expected to be
#> `"bootstrap"`. This workflow is untested and the results may not mean what you
#> think they mean.
#> Error in sample.int(length(x), size, replace, prob): NA in probability vector

# expected bootstrap, given draw (no hypothesis)
gss %>%
  specify(response = hours) %>%
  # hypothesize(null = "point", mu = 40) %>%
  generate(reps = 200, type = "draw")
#> Response: hours (numeric)
#> # A tibble: 100,000 × 2
#> # Groups:   replicate [200]
#>    hours replicate
#>    <fct> <fct>    
#>  1 42    1        
#>  2 40    1        
#>  3 56    1        
#>  4 5     1        
#>  5 30    1        
#>  6 64    1        
#>  7 72    1        
#>  8 78    1        
#>  9 5     1        
#> 10 6     1        
#> # … with 99,990 more rows

# expected and given "draw"
gss %>%
  specify(response = finrela) %>%
  hypothesize(null = "point",
              p = c("far below average" = 1/6,
                    "below average" = 1/6,
                    "average" = 1/6,
                    "above average" = 1/6,
                    "far above average" = 1/6,
                    "DK" = 1/6)) %>%
  generate(reps = 200, type = "draw") %>%
  calculate(stat = "Chisq")
#> Response: finrela (factor)
#> Null Hypothesis: point
#> # A tibble: 200 × 2
#>    replicate  stat
#>    <fct>     <dbl>
#>  1 1          2.75
#>  2 2          4.31
#>  3 3          1.94
#>  4 4          2.70
#>  5 5          3.04
#>  6 6          5.2 
#>  7 7          2.22
#>  8 8         14.5 
#>  9 9          3.4 
#> 10 10         3.81
#> # … with 190 more rows

# expected "draw", given (equivalent) "simulate" (hypothesized)
gss %>%
  specify(response = finrela) %>%
  hypothesize(null = "point",
              p = c("far below average" = 1/6,
                    "below average" = 1/6,
                    "average" = 1/6,
                    "above average" = 1/6,
                    "far above average" = 1/6,
                    "DK" = 1/6)) %>%
  generate(reps = 200, type = "simulate") %>%
  calculate(stat = "Chisq")
#> The `"simulate"` generation type has been renamed to `"draw"`. Use `type = "draw"` instead to quiet this message.
#> Response: finrela (factor)
#> Null Hypothesis: point
#> # A tibble: 200 × 2
#>    replicate  stat
#>    <fct>     <dbl>
#>  1 1          3.83
#>  2 2          6.86
#>  3 3          4.29
#>  4 4         14.4 
#>  5 5          2.63
#>  6 6          4.48
#>  7 7          4   
#>  8 8          4.05
#>  9 9          3.57
#> 10 10         4.96
#> # … with 190 more rows

# expected "draw", given (equivalent) "simulate" (no hypothesis)
gss %>%
  specify(response = finrela) %>%
  generate(reps = 200, type = "simulate") %>%
  calculate(stat = "Chisq")
#> The `"simulate"` generation type has been renamed to `"draw"`. Use `type = "draw"` instead to quiet this message.
#> Warning: A chi-square statistic requires a null hypothesis to calculate the observed statistic. 
#> Output assumes the following null values: `p = c(above average = 0.166666666666667, average = 0.166666666666667, below average = 0.166666666666667, DK = 0.166666666666667, far above average = 0.166666666666667, far below average = 0.166666666666667)`.
#> Response: finrela (factor)
#> Null Hypothesis: point
#> # A tibble: 200 × 2
#>    replicate  stat
#>    <fct>     <dbl>
#>  1 1         2.03 
#>  2 2         2.87 
#>  3 3         0.448
#>  4 4         4.46 
#>  5 5         5.87 
#>  6 6         3.50 
#>  7 7         9.42 
#>  8 8         0.736
#>  9 9         1.31 
#> 10 10        1.12 
#> # … with 190 more rows

# expected draw, given bootstrap (hypothesized)
gss %>%
  specify(response = finrela) %>%
  hypothesize(null = "point",
              p = c("far below average" = 1/6,
                    "below average" = 1/6,
                    "average" = 1/6,
                    "above average" = 1/6,
                    "far above average" = 1/6,
                    "DK" = 1/6)) %>%
  generate(reps = 200, type = "bootstrap") %>%
  calculate(stat = "Chisq")
#> Warning: You have given `type = "bootstrap"`, but `type` is expected to be
#> `"draw"`. This workflow is untested and the results may not mean what you think
#> they mean.
#> Response: finrela (factor)
#> Null Hypothesis: point
#> # A tibble: 200 × 2
#>    replicate  stat
#>        <int> <dbl>
#>  1         1  497.
#>  2         2  489.
#>  3         3  450.
#>  4         4  422.
#>  5         5  444.
#>  6         6  507.
#>  7         7  487.
#>  8         8  472.
#>  9         9  442.
#> 10        10  513.
#> # … with 190 more rows


# expected draw, given bootstrap (no hypothesis)
gss %>%
  specify(response = finrela) %>%
  generate(reps = 200, type = "bootstrap") %>%
  calculate(stat = "Chisq")
#> Warning: A chi-square statistic requires a null hypothesis to calculate the observed statistic. 
#> Output assumes the following null values: `p = c(far below average = 0.166666666666667, below average = 0.166666666666667, average = 0.166666666666667, above average = 0.166666666666667, far above average = 0.166666666666667, DK = 0.166666666666667)`.
#> Response: finrela (factor)
#> Null Hypothesis: point
#> # A tibble: 200 × 2
#>    replicate  stat
#>        <int> <dbl>
#>  1         1  465.
#>  2         2  524.
#>  3         3  476.
#>  4         4  509.
#>  5         5  518.
#>  6         6  551.
#>  7         7  498.
#>  8         8  506.
#>  9         9  539.
#> 10        10  458.
#> # … with 190 more rows
```

<sup>Created on 2022-01-17 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>